### PR TITLE
chore: rename attachment name for lettersg

### PR DESCRIPTION
**Pull request by m0nggh**
## Problem

Alexis mentioned that some users want/hope their template pdf is named more "legitly" i.e. using the template name instead of `letter.pdf`

## Solution

Fetch the template name using the template id chosen in the `$.step.parameters` and replace the name when putting it in the s3 bucket.

Pipes still using the old filename will not be affected because we `getObject` based on the filename and the old files still exist.

Did simple check that no postman body contains `letter.pdf` references!

## Tests
- [x] Old pipes with the lettersg attachment still work
- [x] New pipes with the new attachment name work

### CODE REVIEW

File: packages/backend/src/apps/lettersg/helpers/attachment.ts [hunk 1]
```diff
@@ -20,8 +21,8 @@ export async function downloadAndStoreAttachmentInS3(
     },
   })
 
-  // objectKey: `executionId/appKey/publicId/letter.pdf`
-  const objectKey = `${$.execution.id}/${$.step.appKey}/${publicId}/letter.pdf`
+  // objectKey: `executionId/appKey/publicId/templateName.pdf`
```
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/helpers/attachment.ts]: nit: not sure if this comment adds much value? to the codebase
```suggestion
```

**Review by liangyuanruo — APPROVED** [review]

**Review by pregnantboy — CHANGES_REQUESTED** [review]
can make sure the template name is sanitised? dont want another path traversal vuln

**Review by m0nggh — COMMENTED** [review]
Added copy change after discussing with Alexis because users could be unaware that a letter link would be provided
image

Path traversal is checked when trying to add to s3 bucket so I did not include a check explicitly, and replaced `/` with `_` so that the template name is preserved.

File: packages/backend/src/apps/lettersg/actions/create-letter/index.ts [hunk 2]
```diff
@@ -115,9 +125,21 @@ const action: IRawAction = {
...
       }
 
+      const { data: templateData } = await $.http.get(
+        '/v1/templates/:templateId',
+        {
+          urlPathParams: {
+            templateId: $.step.parameters.templateId,
+          },
+        },
+      )
```
Comment (pregnantboy) [inline packages/backend/src/apps/lettersg/actions/create-letter/index.ts]: nit: this API is called 3 times in different files, could you refactor this with proper handling into a common function

File: packages/backend/src/apps/lettersg/actions/create-letter/index.ts [hunk 3]
```diff
@@ -115,9 +125,21 @@ const action: IRawAction = {
...
+
+      // Note: s3 won't allow for template names with .., we only need to replace / with _ because of how we denote a S3 ID
+      const templateName = templateData.name.replaceAll('/', '_')
```
Comment (pregnantboy) [inline packages/backend/src/apps/lettersg/actions/create-letter/index.ts:132]: lgtm, i dont think we would hit the 1024 max length for object keys

**Review by pregnantboy — APPROVED** [review]
lgtm, other than nit

File: packages/backend/src/apps/lettersg/actions/create-letter/index.ts (hunk 2 above)
Comment (m0nggh) [reply packages/backend/src/apps/lettersg/actions/create-letter/index.ts]: ok got it

**Review by m0nggh — COMMENTED** [review]
new copy change after discussing with Alexis once more
Screenshot 2024-09-20 at 5 21 23 PM

**Pull request merged**